### PR TITLE
Added recoding script for embedded timestamps

### PR DESCRIPTION
**Pull request by begr-gomspace**

### CODE REVIEW

File: full_recode_kleos.py [hunk 1]
```diff
@@ -0,0 +1,326 @@
...
+    def __init__(self):
+        self._crc = mkCrcFun("modbus")
+        self._logger = logging.getLogger("BlockParser")
+        self._sample_chunks = []
+        self._lost_samples = []
+        self._timestamps = np.zeros((0, 2), dtype=np.int64)
+        self._leftover = np.zeros((0, 2), dtype=np.int64)
+        self._last_ts = (0, 0, 0)
+        self._last_sample_index = 0
+        self._first_sample = None
```
Comment (gs-jgj) [inline full_recode_kleos.py:27]:
```suggestion
```

File: full_recode_kleos.py [hunk 2]
```diff
@@ -0,0 +1,326 @@
...
+                sample_index = block_count * 128
+                pps_offset = sample_index - pps_offset  # Transform to absolute offset
+
+                new_ts = (pps_ts, pps_offset)
+                if self._last_ts[0] != new_ts[0]:
+                    self._timestamps = np.concatenate((self._timestamps, np.array([new_ts])))
+                self._last_ts = new_ts
+
+                if self._first_sample is None:
+                    self._first_sample = sample_index
```
Comment (gs-jgj) [inline full_recode_kleos.py:92]:
```suggestion
```

File: full_recode_kleos.py [hunk 3]
```diff
@@ -0,0 +1,326 @@
...
+        else:
+            return None
+
+    def reassemble(self, input_samples):
+        sample_array = np.array(self._sample_chunks, dtype=np.int64)
+
+        min_sample = np.min(sample_array[:, 0])
+        max_sample = np.max(sample_array[:, 0]) + 128
+
+        lost_offsets = np.array(self._lost_samples, dtype=np.int64) - self._first_sample
```
Comment (gs-jgj) [inline full_recode_kleos.py:123]:
```suggestion
        lost_offsets = np.array(self._lost_samples, dtype=np.int64) - min_sample
```

File: full_recode_kleos.py [hunk 4]
```diff
@@ -0,0 +1,326 @@
...
+
+        for index, offset in sample_array:
+            samples[index - min_sample : index - min_sample + 128] = input_samples[
+                offset : offset + 128
+            ]
+
+        self._sample_chunks = []
+        self._lost_samples = []
+
+        return (samples, lost_offsets, self._timestamps, self._first_sample)
```
Comment (gs-jgj) [inline full_recode_kleos.py:135]:
```suggestion

        ts = np.array(self._timestamps, dtype=np.int64)
        ts[:,1] -= min_sample
        ts = ts[ts[:,1] >= 0]

        return (samples, lost_offsets, ts)
```

**Review by gs-jgj — COMMENTED** [review]
Looks good. I made some suggestions regarding how the transformation of timestamps and samples are done just to have that done in one go in the blockparser. Also, how to calculate the pps_offset. The timestamps now align with what would be the expected output.

**Comment by begr-gomspace** [comment]
Total packages + lost at end of program

File: full_recode_kleos.py [hunk 5]
```diff
@@ -0,0 +1,326 @@
...
+                    )
+                    continue
+
+                sample_index = block_count * 128
+                pps_offset = sample_index - pps_offset  # Transform to absolute offset
+
+                new_ts = (pps_ts, pps_offset)
+                if self._last_ts[0] != new_ts[0]:
+                    self._timestamps = np.concatenate((self._timestamps, np.array([new_ts])))
+                self._last_ts = new_ts
```
[the remark this answers is not in the document]
Comment (begr-gomspace) [reply full_recode_kleos.py:89]: Done

File: full_recode_kleos.py (hunk 2 above)
Comment (begr-gomspace) [reply full_recode_kleos.py:92]: Done

File: full_recode_kleos.py (hunk 3 above)
Comment (begr-gomspace) [reply full_recode_kleos.py:123]: Done

File: full_recode_kleos.py (hunk 4 above)
Comment (begr-gomspace) [reply full_recode_kleos.py:135]: Done

File: full_recode_kleos.py [hunk 6]
```diff
@@ -0,0 +1,326 @@
...
+
+        self._local_sample_indices = local_sample_indices
+
+        self._block_count = 0
+        self._sample_index = 0
+        self._timestamps = np.zeros((0, 2), dtype=np.int64)
+        self._buffer = list()
+        self._buffer_offset = 0
+        self._buffer_count = 0
+        self._first_sample = 0
```
[the remark this answers is not in the document]
Comment (begr-gomspace) [reply full_recode_kleos.py:178]: Done

File: full_recode_kleos.py [hunk 7]
```diff
@@ -0,0 +1,326 @@
...
+
+    def _write_bytes(self, byte_count: int, debug=False):
+        sample_count = byte_count // 4
+        # Get timestamps
+        indices = np.where(
+            (self._timestamps[:, 1] >= self._sample_index + self._first_sample)
+            & (self._timestamps[:, 1] < self._sample_index + self._first_sample + sample_count)
```
[the remark this answers is not in the document]
Comment (begr-gomspace) [reply full_recode_kleos.py:185]: Done

File: full_recode_kleos.py [hunk 8]
```diff
@@ -0,0 +1,326 @@
...
+                byte_count = 0
+        timestamps = timestamps[:, 0:2]
+        self._output_file.write(bytes(timestamps.view(np.uint8)))
+
+        # Update counters
+        self._sample_index += sample_count
+        self._block_count += 1
+
+    def write_samples(self, samples: bytes, timestamps, first_sample, debug=False):
+        self._first_sample = first_sample
```
[the remark this answers is not in the document]
Comment (begr-gomspace) [reply full_recode_kleos.py:236]: Done

File: full_recode_kleos.py [hunk 9]
```diff
@@ -0,0 +1,326 @@
...
+    icdr = ICDReader()
+    icdw = ICDWriter(args.packet_size, args.output_file, True)
+    samples = np.zeros((0,2), dtype=np.int16)
+    for s, t in icdr.decodefile(args.input_file):
+        samples = np.concatenate((samples, s))
+    reassembled = bp.parse_samples(samples)
+    
+    newsamples, lost_offsets, timestamp, firstsample = reassembled
+
+    icdw.write_samples(bytes(newsamples.view(np.uint8)), timestamp, firstsample, args.verbose)
```
[the remark this answers is not in the document]
Comment (begr-gomspace) [reply full_recode_kleos.py:262]: Done

File: full_recode_kleos.py [hunk 10]
```diff
@@ -0,0 +1,326 @@
...
+    for lo in lost_offsets:
+        lost_s.append(lo)
+    icdw.flush()
+    with open(args.output_file_missing_indecencies, 'w', newline='') as f:
+        lostoffsetwriter = csv.writer(f, delimiter=',', quotechar='"', quoting=csv.QUOTE_MINIMAL)
+        for lo in lost_s:
+            lostoffsetwriter.writerow(lo)
+        f.flush()
```
[the remark this answers is not in the document]
Comment (begr-gomspace) [reply full_recode_kleos.py:270]: Done

**Review by gs-jgj — APPROVED** [review]

**Pull request merged**